bibblio.py

```diff
@@ -0,0 +1,377 @@
+        qu = self._db.query(Identifier)\
+            .join(Identifier.primarily_identifies).join(Edition.work)\
+            .join(Edition.custom_list_entries).join(CustomListEntry.customlist)\
+            .join(edition_subquery, Edition.id == edition_subquery.c.id)\
+            .filter(CustomList.id==self.custom_list.id)
+
+        if not self.fiction:
+            # Only get nonfiction. This is the default setting.
+            qu = qu.filter(Work.fiction==False)
+
+        if identifiers:
+            qu = qu.filter(Id.id.in_([i.id for i in identifiers]))
+
+        return qu
+
+    def process_item(self, identifier, force=False):
+        content_item = self.content_item_from_identifier(identifier)
+
+        try:
+            result = self.api.create_content_item(content_item)
+        except Exception as e:
+            return CoverageFailure(
+                identifier, str(e), data_source=self.output_source,
+                transient=True
+            )
+
+        content_item_id = result.get('contentItemId')
+        bibblio_identifier, _is_new = Identifier.for_foreign_id(
+            self._db, Identifier.BIBBLIO_CONTENT_ITEM_ID, content_item_id
+        )
+
+        identifier.equivalent_to(
+            self.output_source, bibblio_identifier, 1
+        )
+        return identifier
+
+    def add_coverage_record_for(self, item):
+        """Adds an appropriate CoverageRecord for this particular
+        identifier and any equivalent identifiers.
+
+        Because Bibblio is text-based recommendation source, it doesn't
+        make sense to represent the same work multiple times.
+        """
+        equivalent_identifier_ids = item.equivalent_identifier_ids()[item.id]
+
+        equivalent_identifiers = self._db.query(Identifier).filter(
+            Identifier.id.in_(equivalent_identifier_ids)
+        )
+        for identifier in equivalent_identifiers:
+            super(BibblioCoverageProvider, self).add_coverage_record_for(identifier)
+
+    def content_item_from_identifier(self, identifier):
+        edition = identifier.primarily_identifies[0]
+
+        name = edition.title + ' by ' + edition.author
+        url = self.edition_permalink(edition)
+        text, data_source = self.get_full_text(edition)
+        provider = dict(name=data_source.name)
+
+        content_item = dict(
+            name=name, url=url, text=text, provider=provider
+        )
+        if self.catalogue_id:
+            content_item['catalogueId'] = self.catalogue_id
+
+        return content_item
+
+    @classmethod
+    def edition_permalink(cls, edition):
+        base_url = Configuration.integration_url(
+            Configuration.CONTENT_SERVER_INTEGRATION, required=True
+        )
+        scheme, host = urlparse(base_url)[0:2]
+        base_url = '://'.join([scheme, host])
+
+        urn = edition.primary_identifier.urn
+        permalink = '%s/lookup?urn=%s' % (base_url, urn)
+        return permalink
```

Comment:
Is there any way to use `url_for` here? I know that requires an application context but this is :|